sqldelight-compiler/src/test/kotlin/app/cash/sqldelight/core/queries/ExpressionTest.kt

```diff
@@ -376,6 +378,79 @@ class ExpressionTest {
     ).inOrder()
   }
 
+  @Test fun `coalesce returns the kotlin type for given homogeneous argument types`(dialect: TestDialect) {
+    val file = FixtureCompiler.parseSql(
+      """
+      |import com.example.Foo;
+      |
+      |CREATE TABLE test (
+      |  foo INTEGER AS Foo NOT NULL,
+      |  bar INTEGER AS Foo
+      |);
+      |
+      |someSelect:
+      |SELECT coalesce(foo, bar),
+      |       coalesce(bar, bar),
+      |       coalesce(foo, bar)
+      |FROM test;
+      """.trimMargin(),
+      tempFolder,
+      dialect = dialect.dialect,
+    )
+
+    val query = file.namedQueries.first()
+    assertThat(query.resultColumns.map { it.javaType }).containsExactly(
+      ClassName("com.example", "Foo"),
+      ClassName("com.example", "Foo").copy(nullable = true),
+      ClassName("com.example", "Foo"),
+    ).inOrder()
+  }
+
+  @Test fun `coalesce returns the dialect type for given heterogeneous argument types`(dialect: TestDialect) {
+    val dialectTextType = when {
+      dialect.dialect.isSqlite -> "TEXT"
+      else -> "VARCHAR(1)"
+    }
+
+    val file = FixtureCompiler.parseSql(
+      """
+      |import com.example.Foo;
+      |import com.example.Bar;
+      |
+      |CREATE TABLE test (
+      |  foo INTEGER AS Foo NOT NULL,
+      |  foo2 REAL AS Foo NOT NULL,
+      |  bar INTEGER AS Bar,
+      |  baz $dialectTextType
+      |);
+      |
+      |someSelect:
+      |SELECT coalesce(foo, bar),
+      |       coalesce(foo, foo2),
+      |       coalesce(foo, baz),
+      |       coalesce(bar, baz),
+      |       coalesce(foo, bar, baz)
+      |FROM test;
+      """.trimMargin(),
+      tempFolder,
+      dialect = dialect.dialect,
+    )
+
+    val integerKotlinType = when (dialect) {
+      POSTGRESQL, HSQL, MYSQL -> INT
+      else -> LONG
+    }
+
+    val query = file.namedQueries.first()
+    assertThat(query.resultColumns.map { it.javaType }).containsExactly(
+      integerKotlinType,
+      DOUBLE,
+      STRING,
```

Comment:
Shouldn't be this Foo too?

Reply:
`coalesce(foo, baz),` foo is `INTEGER AS Foo` & `baz` is `TEXT/VARCHAR(1)`, so I'd say no.

Reply:
Hm, I would expect these results (with `TEXT => kotlin.String` as fallback):

 `coalesce(foo INTEGER AS Foo, bar INTEGER AS Bar)` => `kotlin.Int`.
`coalesce(foo INTEGER AS Foo, foo2 REAL AS FOO)` => `Foo`
`coalesce(foo INTEGER AS FOO, baz TEXT)` => `kotlin.String`
`coalesce(bar INTEGER AS BAR, baz TEXT`) => `kotlin.String`
`coalesce(foo INTEGER AS FOO, bar INTEGER AS Bar, baz TEXT)` => `kotlin.String`

Reply:
BTW with the given nullability (`NOT NULL`), we could also assume the output, but I am unsure, if we should do it:

`coalesce(foo INTEGER NOT NULL AS Foo, bar INTEGER AS Bar)` => `Foo`.
`coalesce(foo INTEGER NOT NULL AS Foo, foo2 REAL NOT NULL AS FOO)` => `Foo`
`coalesce(foo INTEGER NOT NULL AS FOO, baz TEXT)` => `Foo`
`coalesce(bar INTEGER AS BAR, baz TEXT)` => `kotlin.String`
`coalesce(foo INTEGER NOT NULL AS FOO, bar INTEGER AS Bar, baz TEXT)` => `Foo`

Reply:
Sorry, after thinking again, you are right. Using different sql types but same kotlin type won't work due to different adapters. So yes, the implementation looks fine to me.